codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1049 @@
-            if (childChannel == null) {
-                // TODO: Combine with DefaultHttp2ConnectionDecoder.shouldIgnoreHeadersOrDataFrame logic.
-                ReferenceCountUtil.release(msg);
-                throw new StreamException(streamId, STREAM_CLOSED, format("Received %s frame for an unknown stream %d",
-                                                                          frame.name(), streamId));
-            }
-            fireChildReadAndRegister(childChannel, frame);
-        } else if (msg instanceof Http2GoAwayFrame) {
-            Http2GoAwayFrame goAwayFrame = (Http2GoAwayFrame) msg;
-            for (PrimitiveEntry<Http2StreamChannel> entry : childChannels.entries()) {
-                Http2StreamChannel childChannel = entry.value();
-                int streamId = entry.key();
-                if (streamId > goAwayFrame.lastStreamId() && isOutboundStream(server, streamId)) {
-                    childChannel.pipeline().fireUserEventTriggered(goAwayFrame.retainedDuplicate());
+    }
+
+    @Override
+    final void onHttp2StreamStateChanged(ChannelHandlerContext ctx, Http2FrameStream stream) {
+        switch (stream.state()) {
+            case OPEN:
+                DefaultHttp2StreamChannel childChannel = new DefaultHttp2StreamChannel(stream);
+                ChannelFuture future = ctx.channel().eventLoop().register(childChannel);
+                if (future.isDone()) {
+                    registerDone(future);
+                } else {
+                    future.addListener(CHILD_CHANNEL_REGISTRATION_LISTENER);
                 }
-            }
-            goAwayFrame.release();
-        } else {
-            // It's safe to release, as UnsupportedMessageTypeException just calls msg.getClass()
-            ReferenceCountUtil.release(msg);
-            throw new UnsupportedMessageTypeException(msg);
+                break;
+            case CLOSED:
+                DefaultHttp2StreamChannel channel = ((Http2MultiplexCodecStream) stream).channel;
+                if (channel != null) {
+                    channel.streamClosed();
+                }
+                break;
+            default:
+                // ignore for now
+                break;
         }
     }
 
-    private void fireChildReadAndRegister(Http2StreamChannel childChannel, Http2StreamFrame frame) {
-        // Can't use childChannel.fireChannelRead() as it would fire independent of whether
-        // channel.read() had been called.
-        childChannel.fireChildRead(frame);
-        if (!childChannel.inStreamsToFireChildReadComplete) {
-            channelsToFireChildReadComplete.add(childChannel);
-            childChannel.inStreamsToFireChildReadComplete = true;
+    @Override
+    final void onHttp2StreamWritabilityChanged(ChannelHandlerContext ctx, Http2FrameStream stream, boolean writable) {
+        DefaultHttp2StreamChannel childChannel = ((Http2MultiplexCodecStream) stream).channel;
+
+        if (childChannel != null)  {
```

Comment:
can you add a comment as to when this is expected to be `null`. Maybe with some of my other suggestions related to writing headers and initializing the stream state ... this won't ever be expected?